File: native/ios/XoteNative/Sources/XoteCommand.swift
import Foundation

/// One entry of a batch, decoded from the wire format in `native/host/protocol.mjs`.
///
/// The wire format is a JSON array of arrays with the opcode first, which is
/// why this is `JSONSerialization` and a switch rather than `Codable`: the
/// elements are deliberately heterogeneous, and a style prop is an arbitrary
/// object.
enum XoteCommand {
  case create(id: Int, type: String)
  case createText(id: Int, text: String)
  case setProp(id: Int, key: String, value: Any?)
  case setText(id: Int, text: String)
  case insert(parent: Int, child: Int, index: Int)
  case remove(parent: Int, child: Int)
  case destroy(id: Int)
  case listen(id: Int, event: String)

  /// How many slots a command of each opcode occupies, opcode included.
  private static func arity(of op: Int) -> Int {
    switch op {
    case 1, 2, 4, 8: return 3
    case 3, 5: return 4
    case 6: return 3
    case 7: return 2
    default: return Int.max
    }
  }

  /// Decode a batch, skipping anything that cannot be read rather than
  /// discarding the whole thing.
  ///
  /// The two halves of this bridge are versioned separately — a JavaScript
  /// bundle can be newer than the app around it — so an opcode this host does
  /// not know is a thing to skip and report, not a reason to drop every other
  /// command in the batch alongside it.
  static func decodeBatch(_ json: String) -> (commands: [XoteCommand], problems: [String]) {
    guard
      let data = json.data(using: .utf8),
      let raw = try? JSONSerialization.jsonObject(with: data) as? [[Any]]
    else {
      return ([], ["a batch arrived that is not an array of commands"])
    }

    var commands: [XoteCommand] = []
    var problems: [String] = []
    for entry in raw {
      if let command = decode(entry) {
        commands.append(command)
      } else {
        problems.append("skipped a command this host cannot read: \(entry)")
      }
    }
    return (commands, problems)
  }

  private static func decode(_ command: [Any]) -> XoteCommand? {
    // Every access below is guarded by the arity check: an index out of range
    // is a trap, not an error, and a truncated batch would take the app down.
    guard let op = command.first as? Int, command.count >= arity(of: op) else { return nil }
    switch op {
    case 1:
      guard let id = command[1] as? Int, let type = command[2] as? String else { return nil }
      return .create(id: id, type: type)
    case 2:
      guard let id = command[1] as? Int, let text = command[2] as? String else { return nil }
      return .createText(id: id, text: text)
    case 3:
      guard let id = command[1] as? Int, let key = command[2] as? String else { return nil }
      // `null` clears the prop, and NSNull is how it survives JSONSerialization.
      let value = command[3] is NSNull ? nil : command[3]
      return .setProp(id: id, key: key, value: value)
    case 4:
      guard let id = command[1] as? Int, let text = command[2] as? String else { return nil }
      return .setText(id: id, text: text)
    case 5:
      guard let parent = command[1] as? Int, let child = command[2] as? Int,
        let index = command[3] as? Int
      else { return nil }
      return .insert(parent: parent, child: child, index: index)
    case 6:
      guard let parent = command[1] as? Int, let child = command[2] as? Int else { return nil }
      return .remove(parent: parent, child: child)
    case 7:
      guard let id = command[1] as? Int else { return nil }
      return .destroy(id: id)
    case 8:
      guard let id = command[1] as? Int, let event = command[2] as? String else { return nil }
      return .listen(id: id, event: event)
    default:
      return nil
    }
  }
}

File: packages/xote-native/hosts/android/app/src/main/java/dev/xote/host/XoteCommand.kt
package dev.xote.host

import org.json.JSONArray
import org.json.JSONObject

/**
 * One entry of a batch, decoded from the wire format in
 * `xote-native/src/host/protocol.mjs`.
 *
 * The wire format is a JSON array of arrays with the opcode first, which is why
 * this is `org.json` and a `when` rather than a data-binding library: the
 * elements are deliberately heterogeneous, and a style prop is an arbitrary
 * object.
 */
sealed class XoteCommand {
  data class Create(val id: Int, val type: String) : XoteCommand()

  data class CreateText(val id: Int, val text: String) : XoteCommand()

  data class SetProp(val id: Int, val key: String, val value: Any?) : XoteCommand()

  data class SetText(val id: Int, val text: String) : XoteCommand()

  data class Insert(val parent: Int, val child: Int, val index: Int) : XoteCommand()

  data class Remove(val parent: Int, val child: Int) : XoteCommand()

  data class Destroy(val id: Int) : XoteCommand()

  data class Listen(val id: Int, val event: String) : XoteCommand()

  data class Batch(val commands: List<XoteCommand>, val problems: List<String>)

  companion object {
    /** How many slots a command of each opcode occupies, opcode included. */
    private fun arity(op: Int): Int =
      when (op) {
        1, 2, 4, 8 -> 3
        3, 5 -> 4
        6 -> 3
        7 -> 2
        else -> Int.MAX_VALUE
      }

    /**
     * Decode a batch, skipping anything that cannot be read rather than
     * discarding the whole thing.
     *
     * The two halves of this bridge are versioned separately — a JavaScript
     * bundle can be newer than the app around it — so an opcode this host does
     * not know is a thing to skip and report, not a reason to drop every other
     * command in the batch alongside it. That is what makes the protocol's
     * append-only rule survivable in practice; see `protocol.mjs`.
     */
    fun decodeBatch(json: String): Batch {
      val commands = mutableListOf<XoteCommand>()
      val problems = mutableListOf<String>()

      val raw =
        try {
          JSONArray(json)
        } catch (error: Exception) {
          return Batch(emptyList(), listOf("batch is not JSON — ${error.message}"))
        }

      for (index in 0 until raw.length()) {
        val entry = raw.optJSONArray(index)
        if (entry == null) {
          problems.add("command $index is not an array")
          continue
        }
        val op = entry.optInt(0, -1)
        if (entry.length() < arity(op)) {
          problems.add("command $index has opcode $op and ${entry.length()} slots")
          continue
        }
        val command =
          when (op) {
            1 -> Create(entry.getInt(1), entry.getString(2))
            2 -> CreateText(entry.getInt(1), entry.getString(2))
            3 -> SetProp(entry.getInt(1), entry.getString(2), plain(entry.opt(3)))
            4 -> SetText(entry.getInt(1), entry.getString(2))
            5 -> Insert(entry.getInt(1), entry.getInt(2), entry.getInt(3))
            6 -> Remove(entry.getInt(1), entry.getInt(2))
            7 -> Destroy(entry.getInt(1))
            8 -> Listen(entry.getInt(1), entry.getString(2))
            else -> null
          }
        if (command == null) {
          problems.add("unknown opcode $op at command $index")
          continue
        }
        commands.add(command)
      }
      return Batch(commands, problems)
    }

    /**
     * `org.json` values into ordinary Kotlin ones, so nothing downstream has to
     * know where a style object came from. `JSONObject.NULL` is a *value*, not
     * `null`, and letting it through would make "clear this prop" arrive as an
     * object that is truthy.
     */
    private fun plain(value: Any?): Any? =
      when (value) {
        null, JSONObject.NULL -> null
        is JSONObject -> value.keys().asSequence().associateWith { plain(value.opt(it)) }
        is JSONArray -> (0 until value.length()).map { plain(value.opt(it)) }
        else -> value
      }
  }
}
